Resources/doc/middlewares.rst

```diff
@@ -0,0 +1,111 @@
+
+    <?php
+
+    namespace App\Middleware;
+
+    use Doctrine\DBAL\Driver;
+    use Doctrine\DBAL\Driver\Middleware;
+
+    class PreventRootConnectionMiddleware implements Middleware
+    {
+        public function wrap(Driver $driver): Driver
+        {
+            return new PreventRootConnectionDriver($driver);
+        }
+    }
+
+As you can see in the ``wrap`` method, the principle of a middleware is
+to decorate Doctrine objects with your own objects bearing the logic you
+need. Now, the ``connect`` method of the driver must be decorated in
+``PreventRootConnectionDriver`` to prevent connections with the root user:
+
+.. code-block:: php
+
+    <?php
+
+    namespace App\Middleware;
+
+    use Doctrine\DBAL\Driver\Connection;
+    use Doctrine\DBAL\Driver\Middleware\AbstractDriverMiddleware;
+    use SensitiveParameter;
+
+    final class PreventRootConnectionDriver extends AbstractDriverMiddleware
+    {
+        public function connect(array $params): Connection
+        {
+            if (isset($params['user']) && $params['user'] === 'root') {
+                throw new \LogicException('Connecting to the database with the root user is not allowed.');
+            }
+
+            return parent::connect($params);
+        }
+    }
+
+That's all! Connection with the root user is not possible anymore. Note
+that ``connect`` is not the only method you can decorate in a ``Connection``.
+But thanks to the ``AbstractDriverMiddleware`` default implementation,
+you only need to decorate the methods for which you want to add some logic.
+Too see a more advanced example with a decoration of the ``Statement`` class,
+you can look at the middleware implementation starting in the class
+:class:`Symfony\\Bridge\\Doctrine\\Middleware\\Debug\\Middleware` of the
+Doctrine Bridge. Decorating the ``Result`` class follows the same principle.
+
+The middleware we've just created applies by default to all the connections.
+If your application has several dbal connections, you can limit the middleware
+scope to a subset of connections thanks to the ``AsMiddleware`` PHP attribute.
+Let's limit our middleware to a connection named ``legacy``:
+
+.. code-block:: php
+
+    <?php
+
+    namespace App\Middleware;
+
+    use Doctrine\Bundle\DoctrineBundle\Attribute\AsMiddleware;
+    use Doctrine\DBAL\Driver;
+    use Doctrine\DBAL\Driver\Middleware;
+
+    #[AsMiddleware(connections: ['legacy'])]
+    class PreventRootConnectionMiddleware implements Middleware
+    {
+        public function wrap(Driver $driver): Driver
+        {
+            return new PreventRootConnectionDriver($driver);
+        }
+    }
+
+All the examples presented above assume ``autoconfigure`` is enabled.
+If ``autoconfigure`` is disabled, the ``doctrine.middleware`` tag must be
+added to the middleware. This tag supports a ``connections`` attribute to
+limit the scope of the middleware.
+
+.. note::
+
+    Middleware have been introduced in version 3.2 of ``doctrine/dbal``
```

Comment:
I'd revert to Middlewares here, or switch to "Support for middleware".

We seem to be treating middleware as countable in the DBAL docs, and the discussion around @derrabus 's PR: https://github.com/doctrine/dbal/pull/5180

I don't know if that's correct… I think we might be referring to middleware [services/decorators], when using the plural form. Middleware is a concept that exists outside the DBAL, what was added in 3.2 was decorators used as middleware.

Reply:
@greg0ire I don't know, I've just pushed a [PR to dbal](https://github.com/doctrine/dbal/pull/6033).

Reply:
If we cannot count middleware, the term can't be plural either, so "middleware have"  sounds wrong. I like @greg0ire's suggestion.

```suggestion
    Support for middleware has been introduced in DBAL 3.2
```